virt_lightning/__init__.py

```diff
@@ -87,25 +100,83 @@ def get_free_ipv4(self):
                 self._last_free_ipv4 = interface
                 return interface
 
+    def get_storage_dir(self):
+        xml = self.pool.XMLDesc(0)
+        root = ET.fromstring(xml)
+        disk_source = root.find("./target/path")
+        return disk_source.text
+
+    def create_disk(self, name, size=20, backing_on=None):
+        disk_path = "{path}/{name}.qcow2".format(path=self.get_storage_dir(), name=name)
+        cmd = ["qemu-img", "create", "-f", "qcow2"]
+        if backing_on:
+            backing_disk = "{path}/upstream/{name}.qcow2".format(
+                path=self.get_storage_dir(), name=backing_on
+            )
+            cmd += ["-b", backing_disk]
+        cmd += [disk_path, "{size}G".format(size=size)]
+
+        run_cmd(cmd)
+        return disk_path
+
+    def prepare_meta_data(self, domain):
+        cidata_path = "{path}/{name}-cidata.iso".format(
+            path=self.get_storage_dir(), name=domain.name()
+        )
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            with open(temp_dir + "/user-data", "w") as fd:
+                fd.write("#cloud-config\n")
+                fd.write(yaml.dump(domain.cloud_init, Dumper=yaml.Dumper))
+            with open(temp_dir + "/meta-data", "w") as fd:
+                fd.write(
+                    domain.meta_data.format(
+                        name=domain.name(), ipv4=domain.ipv4.ip, gateway=domain.gateway
+                    )
+                )
+            with open(temp_dir + "/network-config", "w") as fd:
+                fd.write(yaml.dump(domain._network_meta, Dumper=yaml.Dumper))
+
+            run_cmd(
+                [
+                    "genisoimage",
+                    "-output",
+                    cidata_path,
+                    "-volid",
+                    "cidata",
+                    "-joliet",
+                    "-r",
+                    "user-data",
+                    "meta-data",
+                    "network-config",
+                ],
+                cwd=temp_dir,
+            )
+        return cidata_path
+
+    def start(self, domain):
+        meta_data_iso = self.prepare_meta_data(domain)
+        domain.attachDisk(meta_data_iso, device="cdrom", disk_type="raw")
+        domain.dom.create()
+
 
 class LibvirtDomain:
-    def __init__(self, dom):
+    def __init__(self, dom, hv=None):
         self.dom = dom
         self.cloud_init = []
         self.meta_data = (
             "dsmode: local\n" "instance-id: iid-{name}\n" "local-hostname: {name}\n"
         )
         self._username = None
         self.ssh_key = None
-        self.wait_for = []
         self.distro = None
 
-    def new(conn):
+    def new(hv):
```

Comment:
I seem this function will work like Factory and more convenient use code like this:

```
@classmethod
def new(cls, hv):
    dom = hv.conn.defineXML(ET.tostring(root).decode())
    return cls(dom, hv)
```
That code isn't block merge just for note, it might more useful that this code.